modules/planning/tasks/deciders/path_assessment_decider.cc

```diff
@@ -66,5 +95,94 @@ void PathAssessmentDecider::SetPathInfo() {
   return;
 }
 
+bool PathAssessmentDecider::IsGreatlyOffReferenceLine(
+    const PathData& path_data) {
+  constexpr double kOffReferenceLineThreshold = 20.0;
+  auto frenet_path = path_data.frenet_frame_path();
+  for (size_t i = 0; i < frenet_path.size(); ++i) {
+    if (std::fabs(frenet_path[i].l()) > kOffReferenceLineThreshold) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool PathAssessmentDecider::IsGreatlyOffRoad(
+    const ReferenceLineInfo& reference_line_info, const PathData& path_data) {
+  constexpr double kOffRoadThreshold = 10.0;
+  auto frenet_path = path_data.frenet_frame_path();
+  for (size_t i = 0; i < frenet_path.size(); ++i) {
+    double road_left_width = 0.0;
+    double road_right_width = 0.0;
+    if (reference_line_info.reference_line().GetRoadWidth(
+            frenet_path[i].s(), &road_left_width, &road_right_width)) {
+      if (frenet_path[i].l() > road_left_width + kOffRoadThreshold ||
+          frenet_path[i].l() < -road_right_width - kOffRoadThreshold) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+bool PathAssessmentDecider::IsCollidingWithStaticObstacles(
+    const ReferenceLineInfo& reference_line_info, const PathData& path_data) {
+  // Get all obstacles and convert them into frenet-frame polygons.
+  std::vector<Polygon2d> obstacle_polygons;
+  auto indexed_obstacles = reference_line_info.path_decision().obstacles();
+  for (const auto* obstacle : indexed_obstacles.Items()) {
+    // Filter out unrelated obstacles.
+    //  - Must be non-virtual
+    if (obstacle->IsVirtual()) {
+      continue;
+    }
+    //  - Must not be ignore-decision obstacles.
+    if (obstacle->HasLongitudinalDecision() &&
+        obstacle->HasLateralDecision() &&
+        obstacle->IsIgnore()) {
+      continue;
+    }
+    //  - Must be non-static
+    if (!obstacle->IsStatic() ||
+        obstacle->speed() > 0.5) {
+      continue;
+    }
+    // Convert into polygon and save it.
+    const auto obstacle_sl = obstacle->PerceptionSLBoundary();
+    obstacle_polygons.push_back(
+        Polygon2d({Vec2d(obstacle_sl.start_s(), obstacle_sl.start_l()),
+                   Vec2d(obstacle_sl.start_s(), obstacle_sl.end_l()),
+                   Vec2d(obstacle_sl.end_s(), obstacle_sl.end_l()),
+                   Vec2d(obstacle_sl.end_s(), obstacle_sl.start_l())}));
```

Comment:
emplace_back might be more efficient when you need to construct.

Reply:
emplace_back seems to not work here because it is a vector of Polygon2d and each Polygon2d has multiple constructors, such as being constructed by Box2d or constructed by vector of Vec2d.